src/createBaseRouter.js

```diff
@@ -84,19 +94,30 @@ export default function createBaseRouter({
       }
     }
 
-    componentWillReceiveProps(nextProps) {
+    static getDerivedStateFromProps({ match, resolver, matchContext }, state) {
+      if (state.isInitialRender) {
+        return { isInitialRender: false };
+      }
+
       if (
-        nextProps.match !== this.props.match ||
-        nextProps.resolver !== this.props.resolver ||
-        !isEqual(nextProps.matchContext, this.props.matchContext)
+        match !== state.match ||
+        resolver !== state.resolver ||
+        !isEqual(matchContext, state.matchContext)
       ) {
-        this.shouldResolveMatch = true;
+        return {
+          match,
+          resolver,
+          matchContext,
+          iteration: state.iteration + 1,
+        };
```

Comment:
you could use `getSnapshotBeforeUpdate()` and cheat a bit since it is an instance method: `return this.shouldResolveMatch = true`

Reply:
i need to update stuff before `render`, though, so `this.lastIteration === iteration` evaluates as false in `render` and prevents the unnecessary update